### PR TITLE
Add helper method HTMLFormElement::fire_event

### LINKED ISSUE #8777: Refactor event firing in submit method of HTMLFormElement
I left a comment in the code [here](https://github.com/KiChjang/servo/commit/2d263f6a117cd8f9f548b5bd8c49ef3739f12f40#diff-db2031d1f23ecf1a921fc100cfed753dR174) describing how the code is repeated for firing an event to the form element itself. We should extract this duplicated functionality into a proper method.

Code: `components/script/dom/htmlformelement.rs`

NB: Wait for #8747 to land before attempting.

### PR DESCRIPTION
**Pull request by askobara**
Fixes #8777

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9030)

### CODE REVIEW

**Comment by KiChjang** [comment]
-S-awaiting-review +S-needs-code-changes 

Nice work! I've left a comment on your code comment where the language was unclear as to what the function does.

I would also advise you to comment on the issue to claim it next time before attempting to make a PR - this lets other contributors know that you are working on the issue. We don't potentially want to have two or more people working on the same issue now, do we?

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/htmlformelement.rs, line 422 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9030#-K5yENqrOZqwRSX8Fodw:-K5yENqrOZqwRSX8Fodx:-1890290695) ([raw file](https://github.com/servo/servo/blob/484d81a9c2336f34943d19ea7241ec8c1936c356/components/script/dom/htmlformelement.rs#L422)):*
`Create and fire a simple event.`
`If no 'target' is supplied, then the event target is assumed to be the form element itself.`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9030#-:-K5yEuCtzavMQ5xBEsnj:-925872072)*

**Comment by KiChjang** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9030)*

**Comment by askobara** [comment]
@KiChjang thanks for review and comment, I'll take it into account the next time

**Comment by Manishearth** [comment]
Uh, sorry, this is incorrect (the original issue was incorrect). The refactoring that was desired was a generic `EventTarget.fire_simple_event("name")` that follows the steps [here](https://html.spec.whatwg.org/multipage/webappapis.html#fire-a-simple-event), and use it everywhere in the codebase where the spec says "fire a simple event named X". Also add a second `fire_simple_event_params("name", bubbles, cancelable)` thing.

In such a case, most code would boil down to something like `self.upcast::<EventTarget>().fire_simple_event("something")`. Find other places which call Event::new for a trusted event and make those use these two methods.


The main reason we want this refactoring is because it encapsulates away the trustedness of the event, so there are less opportunities to mess that up.

Sorry about the confusion.

**Comment by KiChjang** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmlformelement.rs, line 179 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9030#-K60ucJTHbP5ZpOxscLt:-K60ucJTHbP5ZpOxscLu:-1172133563) ([raw file](https://github.com/servo/servo/blob/2d61599fe12f8482697a0fb4f708a775c440cdfd/components/script/dom/htmlformelement.rs#L179)):*
The new call would look something like `let event = self.upcast().fire_simple_event( /* parameters */);`. The `upcast()` may require an additional type parameter. Try `upcast::<EventTarget>()` if `upcast()` fails to compile.

---

*[components/script/dom/htmlformelement.rs, line 423 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9030#-K60uDPwGVafTbudLmga:-K60uDPwGVafTbudLmgb:-179490419) ([raw file](https://github.com/servo/servo/blob/2d61599fe12f8482697a0fb4f708a775c440cdfd/components/script/dom/htmlformelement.rs#L423)):*
Okay, since we've cleared up the issue, this function would actually need to be living under `eventtarget.rs`, and its name changed into `fire_simple_event`. The `target` parameter will be unnecessary.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9030)*

**Comment by askobara** [comment]
Get it!

**Comment by askobara** [comment]
Could you clarify for me how to get a Window instance from the EventTarget?
The first thing I thought is use upcast/downcast methods something like `self.upcast::<Node>()` and then use `window_from_node` but it's doesn't looks efficient.

**Comment by KiChjang** [comment]
The simplest way I can think of is to add another parameter to `fire_simple_event` to get the `Window` object.

In fact, I think we can do better - instead of accepting a `Window` object, we accept a `GlobalRef` object, since that's what `Event::new()` reallly wants.

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -786,12 +786,12 @@ impl XMLHttpRequest {
...
         assert!(self.ready_state.get() != rs);
         self.ready_state.set(rs);
+
         let global = self.global.root();
-        let event = Event::new(global.r(),
-                               atom!("readystatechange"),
-                               EventBubbles::DoesNotBubble,
-                               EventCancelable::Cancelable);
-        event.fire(self.upcast());
+        self.upcast::<EventTarget>().fire_simple_event_params("readystatechange",
```
Comment (Manishearth) [inline components/script/dom/xmlhttprequest.rs]: We don't want to fire a simple event here, this calls https://dom.spec.whatwg.org/#concept-event-fire .


...which seems to do the same thing, not sure why there are two. But leave this unchanged for now

**Comment by Manishearth** [comment]
Reviewed 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9030)*

**Comment by Manishearth** [comment]
Looks good to me aside from the onreadystatechange one here. r=me with that removed.

**Comment by askobara** [comment]
r=Manishearth

**Comment by Manishearth** [comment]
@bors-servo r+

Thanks!

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9030#-:-K69eDNa92vDQSwKhgRv:-1544677996)*

**Pull request merged**